### PR TITLE
PYTHON-4305 Fix bson size check

### PR DESCRIPTION
**Pull request by ShaneHarvey**
https://jira.mongodb.org/browse/PYTHON-4305

### CODE REVIEW

File: bson/_cbsonmodule.c [hunk 1]
```diff
@@ -2311,7 +2312,8 @@ static PyObject* get_value(PyObject* self, PyObject* name, const char* buffer,
             memcpy(&code_size, buffer + *position, 4);
             code_size = BSON_UINT32_FROM_LE(code_size);
             /* code_w_scope length + code length + code + scope length */
-            if (!code_size || max < code_size || max < 4 + 4 + code_size + 4) {
+            len = 4 + 4 + code_size + 4;
+            if (!code_size || max < code_size || max < len || len < code_size) {
```
Comment (ShaneHarvey) [inline bson/_cbsonmodule.c:2316]: `len < code_size` catches when `len` overflows.

File: bson/_cbsonmodule.c [hunk 2]
```diff
@@ -2329,12 +2331,9 @@ static PyObject* get_value(PyObject* self, PyObject* name, const char* buffer,
...
             memcpy(&scope_size, buffer + *position, 4);
             scope_size = BSON_UINT32_FROM_LE(scope_size);
-            if (scope_size < BSON_MIN_SIZE) {
-                Py_DECREF(code);
-                goto invalid;
-            }
             /* code length + code + scope length + scope */
-            if ((4 + code_size + 4 + scope_size) != c_w_s_size) {
+            len = 4 + 4 + code_size + scope_size;
+            if (scope_size < BSON_MIN_SIZE || len != c_w_s_size || len < scope_size) {
```
Comment (ShaneHarvey) [inline bson/_cbsonmodule.c:2336]: `len < scope_size` catches when `len` overflows.

**Review by blink1073 — APPROVED** [review]
LGTM!

**Comment by carnil** [comment]
This is related to CVE-2024-5629

**Pull request merged**